sky/jobs/utils.py

```diff
@@ -257,7 +257,22 @@ def _handle_legacy_job(job_id: int):
 
         # For jobs with schedule state:
         pid = tasks[0]['controller_pid']
-        if pid is None:
+        if schedule_state == managed_job_state.ManagedJobScheduleState.DONE:
+            if all(task['status'].is_terminal() for task in tasks):
+                # Turns out this job is fine, even though it got pulled by
+                # get_jobs_to_check_status. Either the job was somehow in an
+                # inconsistent state that already resolved itself, or we have
+                # a bug in the query.
+                logger.warning('We expected to find a nonterminal done job '
+                               f'{job_id}, but all tasks are terminal.')
+                continue
+
+            logger.error(f'Job {job_id} has DONE schedule state, but some '
+                         f'tasks are not terminal. Task statuses: '
+                         f'{", ".join(task["status"].value for task in tasks)}')
+            failure_reason = ('Job controller is DONE, but some tasks are not '
+                              'terminal.')
```

Comment:
Is it possible the state of the task is just cancelling? Would it be possible the job is DONE for the schedule state? Why does the `pid is None` case fail to catch this, or in other word, can we just let the `pid is None` to handle it?

We need some comments for why we need this additional check

```suggestion
            failure_reason = (f'Scheduler has set job {job_id} to be DONE, but some tasks are not '
                                          'terminal.')
```

Reply:
> Is it possible the state of the task is just cancelling? Would it be possible the job is DONE for the schedule state?

It should not normally be possible. We should either transition to CANCELLED or FAILED_CONTROLLER in that case. It could happen if the transition to CANCELLED silently fails, which is what I saw.

> Why does the pid is None case fail to catch this, or in other word, can we just let the pid is None to handle it?

This is not for the legacy jobs case. This is for handling inconsistent state that could arise via an unknown bug in the current code. (E.g. the bugs fixed in the linked PRs.) So we will have a pid in that case.

Reply:
`Inconsistent internal state for job` or something like that